File: Grind75/longestPalindrome.js
function longestPalindrome(s) {
  let start = 0;
  let maxLen = 0;

  function expand(left, right) {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      left--;
      right++;
    }
    // loop exits one step too far in each direction
    // so actual palindrome is s[left+1 ... right-1]
    const len = right - left - 1;
    if (len > maxLen) {
      maxLen = len;
      start = left + 1;
    }
  }

  for (let i = 0; i < s.length; i++) {
    expand(i, i);     // odd-length:  center is s[i]
    expand(i, i + 1); // even-length: center is gap between s[i] and s[i+1]
  }

  return s.slice(start, start + maxLen);

}
// Time  O(n²) — n centers × up to n expansion steps each
// Space O(1)  — only tracking indices, no table

File: Grind75/longestPalindrome.py
def longestPalindrome(s: str) -> int:
    length = 0
    unpaired = set()

    for ch in s:
        if ch in unpaired:
            length += 2
            unpaired.remove(ch)
        else:
            unpaired.add(ch)
    
    return length + 1 if unpaired else 0


# Longest Palindromic Substring
def longestPalindrome(s: str) -> str:
    if not s:
        return ""

    start, end = 0, 0  # indices of the longest palindrome found

    def expand(left: int, right: int) -> (int, int):
        """
        Expand around the indices `left` and `right` as the center,
        returning the (inclusive) bounds of the longest palindrome found.
        """
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        # After the loop, left and right are one step beyond the palindrome
        return left + 1, right - 1

    for i in range(len(s)):
        # Odd-length palindrome (centered at i)
        l1, r1 = expand(i, i)
        if (r1 - l1) > (end - start):
            start, end = l1, r1

        # Even-length palindrome (centered between i and i+1)
        l2, r2 = expand(i, i + 1)
        if (r2 - l2) > (end - start):
            start, end = l2, r2

    return s[start:end + 1]


# Time Complexity: O(n^2), since for each of the n centers we may expand up to O(n) steps.
# Space Complexity: O(1) extra space (ignoring the output substring).
